File: app/src/main/java/ua/com/foxminded/locationtrackera/util/Result.java
package ua.com.foxminded.locationtrackera.util;

import org.jetbrains.annotations.NotNull;

public class Result<T> {

    private Result() {
    }

    @NotNull
    @Override
    public String toString() {
        if (this instanceof Result.Success) {
            Result.Success<T> success = (Result.Success<T>) this;
            return "Success[data=" + success.getData().toString() + "]";
        } else if (this instanceof Result.Error) {
            Result.Error<T> error = (Result.Error<T>) this;
            return "Error[exception=" + error.getError().toString() + "]";
        }
        return "";
    }

    public boolean isSuccessful() {
        return this instanceof Result.Success;
    }

    public final static class Success<T> extends Result<T> {
        private final T data;

        public Success(T data) {
            this.data = data;
        }

        public T getData() {
            return this.data;
        }
    }

    public final static class Error<T> extends Result<T> {
        private final Throwable error;

        public Error(Throwable error) {
            this.error = error;
        }

        public Throwable getError() {
            return this.error;
        }
    }
}

File: models/src/main/java/ua/com/foxminded/locationtrackera/models/util/Result.kt
package ua.com.foxminded.locationtrackera.models.util

open class Result<T> private constructor() {

    override fun toString(): String = when(this) {
        is Success<*> -> "Success[data=" + this.data.toString() + "]"
        is Error<*> -> "Error[exception=" + this.error.toString() + "]"
        else -> ""
    }

    val isSuccessful: Boolean
        get() = this is Success<*>

    class Success<T>(val data: T) : Result<T>()

    class Error<T>(val error: Throwable?) : Result<T>()
}